### PR TITLE
Add sabredav plugin to register environment auth for dav requests

### PR DESCRIPTION
**Pull request by icewind1991**
For https://github.com/nextcloud/user_saml/issues/181

Makes environment variable sso (i.e. kerberos) work with webdav requests

### CODE REVIEW

File: appinfo/app.php [hunk 1]
```diff
@@ -70,6 +70,12 @@
 	return;
 }
 
+$app = new \OCA\User_SAML\AppInfo\Application();
+$dispatcher = \OC::$server->getEventDispatcher();
```
Comment (blizzz) [inline appinfo/app.php]: appears unused

**Comment by icewind1991** [comment]
Extended to plugin to allow anonymous option requests to the dav root for the second part of https://github.com/nextcloud/user_saml/issues/181

File: lib/DavPlugin.php [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+				$this->session->set(Auth::DAV_AUTHENTICATED, $this->auth[$uidMapping]);
+				$this->session->set('user_saml.samlUserData', $this->auth);
+			}
+		}
+
+		if ($request->getMethod() === 'OPTIONS' && $request->getPath() === '') {
+			/** @var CorePlugin $corePlugin */
+			$corePlugin = $this->server->getPlugin('core');
+			// setup a fake tree for anonymous access
+			$this->server->tree = new Tree(new Directory());
```
Comment (MorrisJobke) [inline lib/DavPlugin.php]: The `$path` parameter is missing in the `Directory` constructor. 


```
{
    "reqId": "bcN5NFcZCxKyLffINjpO",
    "level": 3,
    "time": "2018-04-25T07:46:25+00:00",
    "remoteAddr": "...",
    "user": "--",
    "app": "PHP",
    "method": "OPTIONS",
    "url": "/remote.php/webdav/",
    "message": "Missing argument 1 for Sabre\\DAV\\FS\\Node::__construct(), called in /var/www/html/apps/user_saml/lib/DavPlugin.php on line 71 and defined at /var/www/html/3rdparty/sabre/dav/lib/DAV/FS/Node.php#31",
    "userAgent": "DavClnt",
    "version": "13.0.1.1"
}
```

Comment (icewind1991) [reply lib/DavPlugin.php]: fixed

**Comment by MorrisJobke** [comment]
@schiessle @icewind1991 @20xx Ping.

**Comment by MorrisJobke** [comment]
Moved to 15.

**Comment by MorrisJobke** [comment]
@schiessle @icewind1991 @rullzer  A review would be nice

File: appinfo/app.php [hunk 3]
```diff
@@ -77,6 +77,9 @@
 	return;
 }
 
+$app = new \OCA\User_SAML\AppInfo\Application();
+$app->registerDavAuth();
```
Comment (rullzer) [inline appinfo/app.php]: We have an info.xml entry to register DAV plugins

**Comment by schiessle** [comment]
I don't have the setup to test this. But if at least one would rebase it to current master, re-test it and can tell that it works I would be happy to merge it.

@icewind1991 could you do that, and maybe address the last remaining comments?

I would like to do a new release of the SAML app this week, would be nice to have this as well.

File: appinfo/app.php (hunk 3 above)
Comment (icewind1991) [reply appinfo/app.php]: those plugins only get loaded after auth, which is to late here

**Comment by icewind1991** [comment]
rebased and re-tested

**Comment by schiessle** [comment]
OK, let's geht this in

**Comment by schiessle** [comment]
Thanks @icewind1991 !

**Pull request merged**